File: node/src/locate.ts
/**
 * Locate the `cmdop-core` binary the client spawns.
 *
 * Single fat-package model (report 01 §5): all 5 prebuilt binaries are baked
 * into `@cmdop/sdk/bin/cmdop-core-<os>-<arch>[.exe]`. There are no per-platform
 * sub-packages and no `optionalDependencies` — the runtime resolver below picks
 * the host's binary from the local `bin/` dir, so we are structurally immune to
 * npm's optionalDependencies lockfile-pruning bug class (#4828).
 *
 * Resolution order:
 *   1. `CMDOP_CORE_BINARY` env override (dev / local tests / offline escape
 *      hatch — point it at a `go build`'d binary).
 *   2. `bin/cmdop-core-${process.platform}-${process.arch}${.exe}` resolved
 *      relative to THIS package's root (found robustly for ESM + CJS).
 *
 * The binary is `chmod`'d +x defensively (npm/pnpm/umask or a CI artifact
 * upload can strip the exec bit) and the Windows `.exe` suffix is honoured.
 */

import { chmodSync, existsSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

/**
 * Supported `${process.platform}-${process.arch}` keys, kept in lockstep with
 * `core/scripts/stage_packages.py` TARGETS. process.arch is already x64/arm64
 * and process.platform is already darwin/linux/win32, so the key maps 1:1 to a
 * baked binary name (`cmdop-core-<key>`).
 */
const SUPPORTED_PLATFORMS = new Set([
  "darwin-arm64",
  "darwin-x64",
  "linux-x64",
  "linux-arm64",
  "win32-x64",
]);

function binaryName(): string {
  const key = `${process.platform}-${process.arch}`;
  const exe = process.platform === "win32" ? ".exe" : "";
  return `cmdop-core-${key}${exe}`;
}

/**
 * The directory of THIS module, resolved for both ESM (`import.meta.url`) and
 * CJS (`__dirname`). Anchoring to the module file — never `process.cwd()` —
 * means the binary is found wherever the package is installed, including
 * workspaces and symlinked installs.
 */
function moduleDir(): string {
  // CJS bundle: tsup emits `__dirname`. ESM bundle: reconstruct from the URL.
  // Guard `import.meta` access so the CJS build (where it is undefined) does
  // not throw — tsup rewrites one branch per format, but be defensive.
  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const metaUrl = (import.meta as any)?.url as string | undefined;
    if (metaUrl) return dirname(fileURLToPath(metaUrl));
  } catch {
    /* not ESM — fall through to __dirname */
  }
  if (typeof __dirname !== "undefined") return __dirname;
  // Last resort: resolve relative to cwd (should never be reached in practice).
  return resolve(".");
}

/**
 * Find the package root (the dir containing `bin/`). The built bundle lives in
 * `dist/`, so the package root is one level up; but be robust to either layout
 * by checking `./bin` and `../bin` relative to the module dir.
 */
function packageBinDir(): string {
  const here = moduleDir();
  const candidates = [
    join(here, "bin"), // module sits at package root (unbundled / tests)
    join(here, "..", "bin"), // module sits in dist/ (the published bundle)
    join(here, "..", "..", "bin"),
  ];
  for (const c of candidates) {
    if (existsSync(c)) return c;
  }
  // Default to the dist/.. layout even if the dir does not yet exist, so the
  // error below names a sensible path.
  return join(here, "..", "bin");
}

export function locateBinary(): string {
  const override = process.env.CMDOP_CORE_BINARY;
  if (override) {
    if (!existsSync(override)) {
      throw new Error(`CMDOP_CORE_BINARY points at a missing file: ${override}`);
    }
    makeExecutable(override);
    return override;
  }

  const key = `${process.platform}-${process.arch}`;
  if (!SUPPORTED_PLATFORMS.has(key)) {
    throw new Error(
      `cmdop-core has no prebuilt binary for ${key}. ` +
        "Set CMDOP_CORE_BINARY to a `go build -o /path/cmdop-core ./cmd/cmdop-core` output.",
    );
  }

  const binPath = join(packageBinDir(), binaryName());
  if (!existsSync(binPath)) {
    throw new Error(
      `cmdop-core binary not found at ${binPath}. ` +
        "Reinstall @cmdop/sdk (the 5 baked binaries ship in its bin/ dir), " +
        "or set CMDOP_CORE_BINARY to a local `go build` output.",
    );
  }
  makeExecutable(binPath);
  return binPath;
}

function makeExecutable(path: string): void {
  if (process.platform === "win32") return;
  try {
    chmodSync(path, 0o755);
  } catch {
    /* read-only store; the bit may already be set */
  }
}

File: python/src/cmdop/_locate.py
"""Locate the ``cmdop-core`` binary the client spawns.

Single fat-wheel model: one ``py3-none-any`` wheel ships ALL 5 prebuilt
binaries as package data in ``cmdop/_bin/cmdop-core-<plat>-<arch>[.exe]``.
The runtime resolver below picks the host's binary from that dir, so a single
wheel installs everywhere and pip needs no platform-tag matching.

Resolution order:

1. ``CMDOP_CORE_BINARY`` env override (dev / local tests / offline escape hatch
   — point it at a ``go build``'d binary).
2. The host's baked binary, ``cmdop/_bin/cmdop-core-<plat>-<arch>[.exe]``,
   where ``<plat>-<arch>`` comes from ``sys.platform`` + ``platform.machine()``
   normalized to the npm/Node vocabulary (darwin/linux/win32 ; x64/arm64).

In all cases the binary is ``chmod``'d ``+x`` defensively (wheel/zip archives
can strip the executable bit) and the Windows ``.exe`` suffix is honoured.
"""

from __future__ import annotations

import os
import platform
import stat
import sys
from importlib.resources import as_file, files

# sys.platform -> npm/Node `process.platform` vocabulary (darwin/linux/win32).
# sys.platform is already "darwin"/"linux"; Windows reports "win32" too, so the
# only real normalization is making sure we never emit anything else.
_PLATFORMS = {"darwin": "darwin", "linux": "linux", "win32": "win32"}

# platform.machine() (lowercased) -> npm `process.arch` vocabulary (x64/arm64).
_ARCHES = {
    "x86_64": "x64",
    "amd64": "x64",
    "arm64": "arm64",
    "aarch64": "arm64",
}


def _host_slug() -> str | None:
    """Return ``<plat>-<arch>`` for this host, or None if unsupported."""
    plat = _PLATFORMS.get(sys.platform)
    arch = _ARCHES.get(platform.machine().lower())
    if plat is None or arch is None:
        return None
    return f"{plat}-{arch}"


def _binary_name() -> str:
    """The host's baked binary filename, e.g. ``cmdop-core-darwin-arm64``."""
    slug = _host_slug()
    exe = ".exe" if sys.platform == "win32" else ""
    return f"cmdop-core-{slug}{exe}"


def _make_executable(path: str) -> None:
    if sys.platform == "win32":
        return
    try:
        mode = os.stat(path).st_mode
        os.chmod(path, mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
    except OSError:
        # Read-only store (nix, some CI caches) — the bit may already be set.
        pass


def locate_binary() -> str:
    """Return an absolute path to a runnable ``cmdop-core`` binary.

    Raises :class:`FileNotFoundError` if neither the override nor a baked binary
    is present, or if the host platform/arch has no prebuilt binary.
    """
    override = os.environ.get("CMDOP_CORE_BINARY")
    if override:
        if not os.path.exists(override):
            raise FileNotFoundError(
                f"CMDOP_CORE_BINARY points at a missing file: {override}"
            )
        _make_executable(override)
        return override

    slug = _host_slug()
    if slug is None:
        raise FileNotFoundError(
            f"cmdop-core has no prebuilt binary for {sys.platform}/"
            f"{platform.machine().lower()}. Set CMDOP_CORE_BINARY to a "
            "`go build -o /path/cmdop-core ./cmd/cmdop-core` output."
        )

    name = _binary_name()
    try:
        resource = files("cmdop._bin").joinpath(name)
        with as_file(resource) as p:
            path = str(p)
        if not os.path.exists(path):
            raise FileNotFoundError(path)
    except (FileNotFoundError, ModuleNotFoundError, OSError) as exc:
        raise FileNotFoundError(
            f"cmdop-core binary not found ({name}). The 5 baked binaries ship "
            "in the wheel under cmdop/_bin/; reinstall cmdop, or set "
            "CMDOP_CORE_BINARY to a `go build -o /path/cmdop-core "
            "./cmd/cmdop-core` output."
        ) from exc

    _make_executable(path)
    return path
